Rubberduck.RegexAssistant/Atom.cs

```diff
@@ -0,0 +1,232 @@
+                    if (specifier.Value.EndsWith("-\\"))
+                    {
+                        // BOOM!
+                        throw new ArgumentException("Character Ranges that have incorrectly escaped characters as target are not allowed");
+                    }
+                    else if (specifier.Value.Length == 1)
+                    {
+                        // fun... we somehow got to grab a single backslash. Pattern is probably broken
+                        // alas for simplicity we just skip the incorrect backslash
+                        // TODO: make a warning from this.. how?? no idea
+                        continue;
+                    }
+                }
+                CharacterSpecifiers.Add(specifier.Value);
+            }
+        }
+
+        public string Description
+        {
+            get
+            {
+                return string.Format(InverseMatching 
+                    ? AssistantResources.AtomDescription_CharacterClass_Inverted 
+                    : AssistantResources.AtomDescription_CharacterClass
+                    , HumanReadableClass());
+            }
+        }
+
+        private string HumanReadableClass()
+        {
+            return string.Join(", ", CharacterSpecifiers); // join last with and?
+        }
+
+        public bool TryMatch(ref string text)
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    class Group : Atom
+    {
+        public static readonly string Pattern = @"(?<!\\)\((?<expression>.*)(?<!\\)\)";
+        private static readonly Regex Matcher = new Regex("^" + Pattern + "$");
+
+        private readonly IRegularExpression subexpression;
+        private readonly string specifier;
+
+        public Group(string specifier) {
+            Match m = Matcher.Match(specifier);
+            if (!m.Success)
+            {
+                throw new ArgumentException("The given specifier does not denote a Group");
+            }
+            subexpression = RegularExpression.Parse(m.Groups["expression"].Value);
+            this.specifier = specifier;
+        }
+
+        public string Description
+        {
+            get
+            {
+                return string.Format(AssistantResources.AtomDescription_Group, specifier) + "\r\n" + subexpression.Description;
+            }
+        }
+
+        public bool TryMatch(ref string text)
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    class Literal : Atom
+    {
+        public static readonly string Pattern = @"\(?:[bB(){}\\\[\]\.+*?\dnftvrdDwWsS]|u[\dA-F]{4}|x[\dA-F]{2}|[0-7]{3})|.";
+        private static readonly Regex Matcher = new Regex("^" + Pattern + "$");
+        private static readonly ISet<char> EscapeLiterals = new HashSet<char>();
+        private readonly string specifier;
+
+        static Literal() {
+            foreach (char escape in new char[]{ '.', '+', '*', '?', '(', ')', '{', '}', '[', ']', '|', '\\' })
```

Comment:
Further micro-optimizing would declare that `char[]` as a `private static readonly` field, so that it's not re-created everytime the method runs :wink: